lib/grizzly/firmware_updates/firmware_update_runner.ex

```diff
@@ -194,6 +199,34 @@ defmodule Grizzly.FirmwareUpdates.FirmwareUpdateRunner do
     :ok
   end
 
+  defp handle_nack_response_or_timeout(%FirmwareUpdate{} = firmware_update, type) do
+    if firmware_update.current_command_attempts > 10 do
+      Logger.error(
+        "[Grizzly] Received #{inspect(type)} while updating firmware of device #{firmware_update.device_id}, fragment #{firmware_update.fragment_index}, attempts so far: #{firmware_update.current_command_attempts}. Giving up."
+      )
+
+      firmware_update
+    else
+      Logger.warning(
+        "[Grizzly] Received #{inspect(type)} while updating firmware of device #{firmware_update.device_id}, fragment #{firmware_update.fragment_index}, attempts so far: #{firmware_update.current_command_attempts}. Retransmitting..."
+      )
+
+      {:ok, command_ref} =
+        AsyncConnection.send_command(firmware_update.conn, firmware_update.current_command,
+          timeout: 120_000,
```

Comment:
Will we get any progress while this command is sending and being received by the device? I'm thinking about how the `progress_timeout` is also 20 minutes and independent from this. So if we got to this point after 30 seconds, we may progress timeout before this command is finished and gives us info back. Or maybe we don't care about that and a hard 2 minute limit for the _any_ progression is what is needed (as coded)?

Reply:
This timeout is how long to wait for an ACK or NACK from zipgateway before considering the command failed. Hitting it should be a rare occurrence (ideally impossible, but their seems to be some weirdness that can happen when an S2 nonce exchange fails due to interference).

In any case, it's definitely possible that we hit this condition, and then 10 seconds later the device sends another Firmware Update Metadata Get for the next batch of fragments and we resume making progress.

Also important to note that this is not a blocking call. The timeout happens in another process and is reported to us async, which ends up in `handle_nack_response_or_timeout/2`.

Reply:
Actually, considering what I just said, I'm going to change how timeouts are handled. By the time the FirmwareUpdateRunner process is notified of a timeout, there are two likely scenarios: (a) we've made progress because the target device resumed the download, in which case we don't want to blindly retry the most recent command or (b) we've already made multiple attempts and the target device either isn't retrying or we can't hear it.